src/Portal.js

```diff
@@ -23,81 +23,40 @@ class Portal extends React.Component {
     container: PropTypes.oneOfType([
       componentOrElement,
       PropTypes.func
-    ])
+    ]),
+
+    onRendered: PropTypes.func,
   };
 
   componentDidMount() {
     this._isMounted = true;
-    this._renderOverlay();
-  }
-
-  componentDidUpdate() {
-    this._renderOverlay();
+    this.setContainer()
+    this.forceUpdate(this.props.onRendered)
   }
 
   componentWillReceiveProps(nextProps) {
-    if (this._overlayTarget && nextProps.container !== this.props.container) {
-      this._portalContainerNode.removeChild(this._overlayTarget);
-      this._portalContainerNode = getContainer(nextProps.container, ownerDocument(this).body);
-      this._portalContainerNode.appendChild(this._overlayTarget);
+    if (nextProps.container !== this.props.container) {
+      this.setContainer(nextProps)
     }
   }
 
   componentWillUnmount() {
     this._isMounted = false;
-    this._unrenderOverlay();
-    this._unmountOverlayTarget();
-  }
-
-
-  _mountOverlayTarget = () => {
-    if (!this._overlayTarget) {
-      this._overlayTarget = document.createElement('div');
-      this._portalContainerNode = getContainer(this.props.container, ownerDocument(this).body);
-      this._portalContainerNode.appendChild(this._overlayTarget);
-    }
-  }
-
-  _unmountOverlayTarget = () => {
-    if (this._overlayTarget) {
-      this._portalContainerNode.removeChild(this._overlayTarget);
-      this._overlayTarget = null;
-    }
     this._portalContainerNode = null;
   }
 
-  _renderOverlay = () => {
-
-    let overlay = !this.props.children
-      ? null
-      : React.Children.only(this.props.children);
-
-    // Save reference for future access.
-    if (overlay !== null) {
-      this._mountOverlayTarget();
-      this._overlayInstance = ReactDOM.unstable_renderSubtreeIntoContainer(
-        this, overlay, this._overlayTarget
-      );
-    } else {
-      // Unrender if the component is null for transitions to null
-      this._unrenderOverlay();
-      this._unmountOverlayTarget();
-    }
-  }
-
-  _unrenderOverlay = () => {
-    if (this._overlayTarget) {
-      ReactDOM.unmountComponentAtNode(this._overlayTarget);
-      this._overlayInstance = null;
-    }
+  setContainer = (props = this.props) => {
+    this._portalContainerNode = getContainer(props.container, ownerDocument(this).body);
```

Comment:
why not just use `setState` here?

Reply:
can we handle setting this in the ctor in cases where it is defined there (for whatever reason?)

Reply:
hmm, is there any meaningful impact to no longer creating a `<div>` here? what if the container does end up being `document`?

Reply:
I don't think i understood what the div was doing in the old code, the container exists somewhere why were we then appending it to a different parent?

Reply:
I'd like an earlier path but it takes some reworking. right now we do `findDOMNode` on the result of the function or the value, and we can't reasonable know when to not do that, or when it will error. I think we need to change the API to provide a DOM node specifically vs a component instance. e.g. `() => findDOMNode(this)`, i think we do that for `target` in Position so it's probably a good idea anyway

Reply:
Maybe just fast-path `container && container.nodeType === Node.ELEMENT_NODE`?

Reply:
I think the `<div>` is like – if the container ends up being `document`, you don't want to try to mount something on `document`.

Reply:
@taion there are too many issues with doing a fast path. For instance, we default `container` to the owner `document`, which we get by doing `findDOMNode` on the component instance (Portal/Modal), in the early case that won't work because the component hasn't mounted yet.

these can can be individually handled, but gets harder across components, e.g Portal mounts early, and calls back to Modal's `onShow` which assumes/needs the owner document

Reply:
Okay, forget the fast path then.

What are your thoughts around adding the `<div>`? Is that just a legacy thing to prevent trying to e.g. unmount the tree rooted at `document`?

Reply:
I'm not sure honestly but it seems safe to remove. if someone wants to try and mount over their application that's sort of their problem :P my guess as to why it's there is so you don't blow away the dom that's already there when you render() but createPortal already appends into it's container

Reply:
okay, so just to confirm, nothing will go wrong with using `document` as the portal node?

Reply:
having the container be `document` wouldn't work before either. if it's document.body or any thing else i'll work fine